### PR TITLE
[ fix #767 ] Format strings and Printf

### LINKED ISSUE #767: Add Text.Printf
Can't believe we don't have this classic example of a dependently-typed primitive! :D

Technical remark: for the catchall-case to reduce in the type of the function, it is important not to define the type by induction on the list of characters but rather on a list of tokens which will have been generated from the list of characters.

i.e. the beginning of the library looks something like this:

```agda
data Chunk : Set where
  Nat : Chunk
  (...)
  Raw : String -> Chunk

toChunks : List Char → List Chunk
toChunks ('%' ∷ 'u' ∷ cs) = Nat ∷ toChunks cs
(...)

-- and then:

Printf : List Chunk → Set
Printf (...)

printf : (cs : List Chunk) → Printf cs
```

### PR DESCRIPTION
**Pull request by gallais**
Beginning the work on printf.

### CODE REVIEW

File: src/Text/Format.agda [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+open import Data.String.Base
+
+------------------------------------------------------------------------
+-- Basic types
+
+data Chunk : Set where
+  `ℕ `ℤ `Float `Char `String : Chunk
+  Raw : String → Chunk
+
+Format = List Chunk
```
Comment (MatthewDaggitt) [inline src/Text/Format.agda]: Minor nitpick, but I'd add a type in for `Format`

**Comment by gallais** [comment]
As is often the case, writing some documentation uncovered some issues (the error
message was not normalised enough which made it hard to read) and we now have a
better PR for it!

**Comment by MatthewDaggitt** [comment]
That's amazing that you get such a neat error message out of it. I'd never have believe it was possible :+1: 
Merging in.

**Comment by erydo** [comment]
This is really awesome, @gallais

**Pull request merged**